common/comms/src/inbound_message_service/message_context.rs

```diff
@@ -0,0 +1,146 @@
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::inbound_message_service::{
+    comms_msg_handlers::determine_comms_msg_dispatch_type,
+    message_dispatcher::Dispatchable,
+};
+use serde_derive::{Deserialize, Serialize};
+use std::convert::TryFrom;
+use tari_comms::connection::message::{FrameSet, MessageEnvelopeHeader, MessageError};
+use tari_crypto::keys::PublicKey;
+
+/// The number of frames required to construct a MessageContext
+const FRAMES_PER_MESSAGE: usize = 5;
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct MessageContext<PubKey> {
+    pub connection_id: Vec<u8>,
+    pub source: Vec<u8>,
+    pub version: Vec<u8>,
+    pub node_identity: Option<PubKey>,
+    pub message_envelope_header: MessageEnvelopeHeader<PubKey>,
+    pub message_envelope_body: Vec<u8>,
+}
+
+impl<PubKey: PublicKey + 'static> MessageContext<PubKey> {
+    /// Construct a new MessageContext that consist of the peer connection information and the received message header
+    /// and body
+    pub fn new(
+        connection_id: Vec<u8>,
+        source: Vec<u8>,
+        version: Vec<u8>,
+        node_identity: Option<PubKey>,
+        message_envelope_header: MessageEnvelopeHeader<PubKey>,
+        message_envelope_body: Vec<u8>,
+    ) -> MessageContext<PubKey>
+    {
+        MessageContext {
+            connection_id,
+            source,
+            version,
+            node_identity,
+            message_envelope_header,
+            message_envelope_body,
+        }
+    }
+
+    /// Serialize the MessageContext into a FrameSet
+    pub fn to_frame_set(&self) -> Result<FrameSet, MessageError> {
+        let mut frame_set: Vec<Vec<u8>> = Vec::new();
+        frame_set.push(self.connection_id.clone());
+        frame_set.push(self.source.clone());
+        frame_set.push(self.version.clone());
+        // node identity should be excluded
+        frame_set.push(self.message_envelope_header.to_frame()?);
+        frame_set.push(self.message_envelope_body.clone());
+        Ok(frame_set)
+    }
+}
+
+impl<PubKey: PublicKey> TryFrom<FrameSet> for MessageContext<PubKey> {
+    type Error = MessageError;
+
+    /// Attempt to create a MessageContext from a FrameSet
+    fn try_from(frames: FrameSet) -> Result<Self, Self::Error> {
+        if frames.len() == FRAMES_PER_MESSAGE {
+            Ok(MessageContext {
+                connection_id: frames[0].clone(),
```

Comment:
Since you grab ownership of `frames`, can't you avoid all these `clones`?

Reply:
To implement the ```try_from``` you cannot pass the value as mutable or by reference so drain and remove doesn't work. I seem to need the clone, otherwise I am trying to move out of the borrowed content as it is a vec.

Reply:
It was possible with remove, change applied